fv/workload/workload.go

```diff
@@ -297,6 +299,177 @@ func (w *Workload) LatencyTo(ip, port string) time.Duration {
+	outputBytes, err := utils.Command("docker", "exec", s.W.C.Name,
+		"cat", s.PidFile).CombinedOutput()
+	if err != nil {
+		return err
+	}
+	pid := strings.TrimSpace(string(outputBytes))
+	err = utils.Command("docker", "exec", s.W.C.Name, "kill", pid).Run()
+	if err != nil {
+		return err
+	}
+	s.RunCmd.Process.Wait()
+	log.WithField("SideService", s).Info("Side service now stopped")
+	return nil
+}
+
+func (w *Workload) StartSideService() *SideService {
+	s, err := startSideService(w)
+	Expect(err).NotTo(HaveOccurred())
+	return s
+}
+
+func startSideService(w *Workload) (*SideService, error) {
+	// Ensure that the host has the 'test-workload' binary.
+	w.C.EnsureBinary("test-workload")
+	sideServIdx++
+	n := fmt.Sprintf("%s-ss%d", w.Name, sideServIdx)
+	pidFile := fmt.Sprintf("/tmp/%s-pid", n)
+
+	testWorkloadShArgs := []string{
+		"/test-workload",
+	}
+	if w.Protocol == "udp" {
+		testWorkloadShArgs = append(testWorkloadShArgs, "--udp")
+	}
+	testWorkloadShArgs = append(testWorkloadShArgs,
+		"--iptables",
+		"--up-lo",
+		fmt.Sprintf("'--namespace-path=%s'", w.namespacePath),
+		"''", // interface name, not important
+		"127.0.0.1",
+		"15001",
+	)
+	pidCmd := fmt.Sprintf("echo $$ >'%s'", pidFile)
+	testWorkloadCmd := strings.Join(testWorkloadShArgs, " ")
+	dockerWorkloadArgs := []string{
+		"docker",
+		"exec",
+		w.C.Name,
+		"sh", "-c",
+		fmt.Sprintf("%s; exec %s", pidCmd, testWorkloadCmd),
+	}
+	runCmd := utils.Command(dockerWorkloadArgs[0], dockerWorkloadArgs[1:]...)
+	logName := fmt.Sprintf("side service %s", n)
+	if err := utils.LogOutput(runCmd, logName); err != nil {
+		return nil, fmt.Errorf("failed to start output logging for %s", logName)
+	}
+	if err := runCmd.Start(); err != nil {
+		return nil, fmt.Errorf("starting /test-workload as a side service failed: %v", err)
+	}
+	return &SideService{
+		W:       w,
+		Name:    n,
+		RunCmd:  runCmd,
+		PidFile: pidFile,
+	}, nil
+}
+
+type PermanentConnection struct {
+	W        *Workload
+	LoopFile string
+	Name     string
+	RunCmd   *exec.Cmd
+}
+
+func (pc *PermanentConnection) Stop() {
+	Expect(pc.stop()).NotTo(HaveOccurred())
+}
+
+func (pc *PermanentConnection) stop() error {
+	if err := utils.Command("docker", "exec", pc.W.C.Name,
+		"sh", "-c", fmt.Sprintf("echo > %s", pc.LoopFile)).Run(); err != nil {
+		return err
+	}
+	if err := pc.RunCmd.Wait(); err != nil {
+		return err
+	}
+	return nil
+}
+
+func (w *Workload) StartPermanentConnection(ip string, port, sourcePort int) *PermanentConnection {
+	pc, err := startPermanentConnection(w, ip, port, sourcePort)
+	Expect(err).NotTo(HaveOccurred())
+	return pc
+}
+
+func startPermanentConnection(w *Workload, ip string, port, sourcePort int) (*PermanentConnection, error) {
+	// Ensure that the host has the 'test-connection' binary.
+	w.C.EnsureBinary("test-connection")
+	permConnIdx++
+	n := fmt.Sprintf("%s-pc%d", w.Name, permConnIdx)
+	loopFile := fmt.Sprintf("/tmp/%s-loop", n)
+
+	err := utils.Command("docker", "exec", w.C.Name,
```

Comment:
We have a utility method for running commands inside the container.  `w.Exec` or `w.C.Exec`, I think

Reply:
Done.